File: examples/server/03_chat/01_basic/java/Main.kt
import com.joinself.selfsdk.kmp.account.Account
import com.joinself.selfsdk.kmp.account.LogLevel
import com.joinself.selfsdk.kmp.account.Target
import com.joinself.selfsdk.kmp.event.Message
import com.joinself.selfsdk.kmp.keypair.signing.PublicKey
import com.joinself.selfsdk.kmp.message.Chat
import com.joinself.selfsdk.kmp.message.ChatBuilder
import com.joinself.selfsdk.kmp.message.ContentType
import java.time.LocalTime
import java.time.format.DateTimeFormatter

fun main() {
    println("💬 Simple Chat Demo")
    println("===================")

    // Create and configure Self account
    println("🔧 Setting up Self account...")
    val account = Common.setupAccount(callbacks = object: Common.Callbacks {
        override fun onMessage(account: Account, msg: Message) {
            handleMessage(account, msg)
        }
    })
    println("✅ Self account created successfully")
    println("✅ Connected to Self network")

    Common.displayAccountInfo(account, "Chat Account")

    println("✅ Chat demo ready! Press Ctrl+C to exit.")

    // Keep running to handle incoming messages
    try {
        Thread.sleep(Long.MAX_VALUE)
    } catch (e: InterruptedException) {
        println("\n👋 Shutting down...")
    }
}

/**
 * Processes all incoming messages
 */
fun handleMessage(account: Account, message: Message) {
    val timestamp = getCurrentTimestamp()
    val type = message.content().contentType()
    when(type) {
        ContentType.CHAT -> {
            handleChatMessage(account, message, timestamp)
        }
        else -> {
            println("📨 [$timestamp] Received message $type from ${message.fromAddress().encodeHex()}")
        }
    }
}

/**
 * Processes incoming chat messages and sends responses
 */
fun handleChatMessage(account: Account, msg: Message, timestamp: String) {
    val chat = Chat.decode(msg.content())
    println("📨 [$timestamp] ${msg.fromAddress().encodeHex()}: ${chat.message()}")

    val response = generateResponse(chat.message(), timestamp)
    sendResponse(account, msg.fromAddress(), response, timestamp)
}

/**
 * Creates appropriate responses based on the message content
 */
fun generateResponse(messageText: String, timestamp: String): String {
    val message = messageText.lowercase().trim()
    
    return when {
        message.contains("hello") || message.contains("hi") -> {
            "👋 Hello! Message received at $timestamp"
        }
        message.contains("how are you") -> {
            "🤖 I'm doing great! I'm a Self SDK chat demo."
        }
        message.contains("help") -> {
            "💡 I'm a simple chat bot. Try saying 'hello', 'how are you', or send any message for an echo!"
        }
        message.contains("time") -> {
            "🕐 Current time is $timestamp"
        }
        else -> {
            "🔄 Echo: $messageText"
        }
    }
}

/**
 * Sends a chat response to the peer
 */
fun sendResponse(account: Account, toAddress: PublicKey, responseText: String, timestamp: String) {
    val chat = ChatBuilder()
        .message(responseText)
        .finish()
    val sendStatus = account.messageSend(toAddress, chat)
    if (sendStatus.success()) {
        println("📤 [$timestamp] Sent: \"$responseText\"\n")
    } else {
        println("❌ [$timestamp] Failed to send response: ${sendStatus.errorMessage()}")
    }
}

/**
 * Helper function to get current timestamp
 */
fun getCurrentTimestamp(): String {
    return LocalTime.now().format(DateTimeFormatter.ofPattern("HH:mm:ss"))
} 

File: examples/server/03_chat/01_basic/go/main.go
// Package main demonstrates simple chat messaging using the Self SDK
package main

import (
	"fmt"
	"strings"
	"time"

	"github.com/joinself/academy/examples/go/common"
	"github.com/joinself/self-go-sdk/account"
	"github.com/joinself/self-go-sdk/event"
	"github.com/joinself/self-go-sdk/keypair/signing"
	"github.com/joinself/self-go-sdk/message"
)

func main() {
	fmt.Println("💬 Simple Chat Demo")
	fmt.Println("===================")

	// Create and configure Self account
	selfAccount := common.SetupAccount(common.AccountConfig{
		Callbacks: account.Callbacks{
			OnMessage: handleMessage,
			OnConnect: func(acc *account.Account) {
				fmt.Println("✅ Connected to Self network")
			},
		},
	})

	// Display connection information
	common.DisplayAccountInfo(selfAccount, "Chat Account")

	fmt.Println("✅ Chat demo ready! Press Ctrl+C to exit.")

	// Keep running to handle incoming messages
	select {}
}

// handleMessage processes all incoming messages
func handleMessage(acc *account.Account, msg *event.Message) {
	timestamp := time.Now().Format("15:04:05")

	switch event.ContentTypeOf(msg) {
	case message.ContentTypeChat:
		handleChatMessage(acc, msg, timestamp)
	default:
		fmt.Printf("📨 [%s] Received message from %s\n",
			timestamp, msg.FromAddress().String())
	}
}

// handleChatMessage processes incoming chat messages and sends responses
func handleChatMessage(acc *account.Account, msg *event.Message, timestamp string) {
	chat, err := message.DecodeChat(msg.Content())
	if err != nil {
		fmt.Printf("❌ [%s] Failed to decode chat message: %v\n", timestamp, err)
		return
	}

	fmt.Printf("📨 [%s] %s: \"%s\"\n", timestamp, msg.FromAddress().String(), chat.Message())

	response := generateResponse(chat.Message(), timestamp)
	sendResponse(acc, msg.FromAddress(), response, timestamp)
}

// generateResponse creates appropriate responses based on the message content
func generateResponse(messageText, timestamp string) string {
	message := strings.ToLower(strings.TrimSpace(messageText))

	switch {
	case strings.Contains(message, "hello") || strings.Contains(message, "hi"):
		return fmt.Sprintf("👋 Hello! Message received at %s", timestamp)
	case strings.Contains(message, "how are you"):
		return "🤖 I'm doing great! I'm a Self SDK chat demo."
	case strings.Contains(message, "help"):
		return "💡 I'm a simple chat bot. Try saying 'hello', 'how are you', or send any message for an echo!"
	case strings.Contains(message, "time"):
		return fmt.Sprintf("🕐 Current time is %s", timestamp)
	default:
		return fmt.Sprintf("🔄 Echo: %s", messageText)
	}
}

// sendResponse sends a chat response to the peer
func sendResponse(acc *account.Account, toAddress *signing.PublicKey, responseText, timestamp string) {
	chatContent, err := message.NewChat().
		Message(responseText).
		Finish()
	if err != nil {
		fmt.Printf("❌ [%s] Failed to build response: %v\n", timestamp, err)
		return
	}

	err = acc.MessageSend(toAddress, chatContent)
	if err != nil {
		fmt.Printf("❌ [%s] Failed to send response: %v\n", timestamp, err)
	} else {
		fmt.Printf("📤 [%s] Sent: \"%s\"\n", timestamp, responseText)
	}
}
